x-pack/plugins/security_solution/public/detections/components/rules/step_define_rule/index.tsx

```diff
@@ -324,6 +337,36 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
       threatIndexPatternsLoading,
     ]
   );
+
+  const onOptionsChange = useCallback((field: FieldsEqlOptions, value: string | undefined) => {
+    setOptionsSelected((prevOptions) => ({
+      ...prevOptions,
+      [field]: value,
+    }));
+  }, []);
+
+  const optionsData = useMemo(
+    () =>
+      isEmpty(indexPatterns.fields)
+        ? {
+            keywordFields: [],
+            dateFields: [],
+            nonDateFields: [],
+          }
+        : {
+            keywordFields: (indexPatterns.fields as FieldSpec[])
```

Comment:
Instead of iterating over the same data three times here, why not reduce? It might not be that much of a performance hit, but we've seen cases where users have 20,000+ fields  🤷🏾 . Either way, this is what I was thinking about
```
indexPatterns.fields.reduce((acc, field) => {
	const fieldLabel = { label: field.name };
	if (field.esTypes?.includes('keyword')) acc.keywordFields.push(fieldLabel);
	if (field.type === 'date') acc.dateFields.push(fieldLabel);
	if (field.type !== 'date') acc.nonDateFields.push(fieldLabel);
	return acc;
}, { keywordFields: [], dateFields: [], nonDateFields: [] });
```

Reply:
Also out of curiosity why not use `esTypes` for the date as well instead of type?